### PR TITLE
Add worker pool for LDAP group lookup

### PR DESCRIPTION
**Pull request by jasonodonnell**
This PR aims to make LDAP group lookup more performant by adding a worker pool for group searches. Specifically this is helpful when `use_token_groups=true` and we need to look up every group individually in LDAP associated with the user.

In my testing where a LDAP user is a member of 300 LDAP groups and has 300 policies mapped to those groups I see significant performance increase by using a worker pool:

Before:
```bash
$ time vault login -method=ldap username=bob password=password
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  hvs.CAESIN2lzMYdtMMQ31k5MoBccnS8exljazJc6ahv40RN9sfJGh4KHGh2cy5kSGFyS3diVXNPRmIwWG5pdkw4OWV2Sjg
token_accessor         rOEMnAWD9hUWMqUL1Md5ko84
token_duration         768h
token_renewable        true
...

real	0m16.168s
user	0m0.057s
sys	0m0.027s
```

After:
```bash
$ time vault login -method=ldap username=bob password=password
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  hvs.CAESIE311YhuFnlcsN2BYJKc0dnOnMNMnXLRBOMxpaXcD1s5Gh4KHGh2cy5iZjc0dkx5ZjdHVFBySUxiUERmTGpMSXg
token_accessor         n8QwGvA5HDXIcfB77FaT3hye
token_duration         768h
token_renewable        true
...

real	0m2.303s
user	0m0.059s
sys	0m0.029s
```

### CODE REVIEW

File: sdk/helper/ldaputil/client.go [hunk 1]
```diff
@@ -498,37 +502,49 @@ func (c *Client) performLdapTokenGroupsSearch(cfg *ConfigEntry, conn Connection,
...
+				if err != nil {
+					c.Logger.Warn("unable to read the group sid", "sid", sid)
+					continue
+				}
+				if len(groupResult.Entries) == 0 {
+					c.Logger.Warn("unable to find the group", "sid", sid)
+					continue
+				}
+
+				groupEntries = append(groupEntries, groupResult.Entries[0])
```
Comment (mpalmi) [inline sdk/helper/ldaputil/client.go:531]: I could be mistaken, but I don't think this is thread-safe. You might need to introduce a lock or something to achieve that, unless you want to have these entries stored in a `map[int]*ldap.Entry` where each key is guaranteed to only be accessed by a single thread.

Also, does ordering matter here? if the loop below is fast enough, we might end up with a non-deterministic order and possibly have to sort at the end...

Comment (mpalmi) [reply sdk/helper/ldaputil/client.go:531]: To that end, we might want to add a unit test for this method to make sure we're testing these things, unless this is already covered by one of the existing tests.

Comment (jasonodonnell) [reply sdk/helper/ldaputil/client.go:534]: You're right, this wasn't thread safe and needed a lock! 😄

Order does not matter. These are used to look up Vault groups later and match them to policies. Vault seems to order these and subsequent logins all look the same:

```bash
$ time vault login -method=ldap username=bob password=password
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  hvs.CAESII1c9Z8LNLTfF5qFfeJTs9QKI9AFqxHdC5ZSLLvWoJ_cGh4KHGh2cy5pdVJTN255V29jaDdCaHlMY0xrWWZ4YjA
token_accessor         PP3Hyvpp6DCQ5FVJRkhc8rfJ
token_duration         768h
token_renewable        true
token_policies         ["default" "dev-1" "dev-10" "dev-100" "dev-101" "dev-102" "dev-103" "dev-104" "dev-105" "dev-106" "dev-107" "dev-108" "dev-109" "dev-11" "dev-110" "dev-111" "dev-112" "dev-113" "dev-114" "dev-115" "dev-116" "dev-117" "dev-118" "dev-119" "dev-12" "dev-120" "dev-121" "dev-122" "dev-123" "dev-124" "dev-125" "dev-126" "dev-127" "dev-128" "dev-129" "dev-13" "dev-130" "dev-131" "dev-132" "dev-133" "dev-134" "dev-135" "dev-136" "dev-137" "dev-138" "dev-139" "dev-14" "dev-140" "dev-141" "dev-142" "dev-143" "dev-144" "dev-145" "dev-146" "dev-147" "dev-148" "dev-149" "dev-15" "dev-150" "dev-151" "dev-152" "dev-153" "dev-154" "dev-155" "dev-156" "dev-157" "dev-158" "dev-159" "dev-16" "dev-160" "dev-161" "dev-162" "dev-163" "dev-164" "dev-165" "dev-166" "dev-167" "dev-168" "dev-169" "dev-17" "dev-170" "dev-171" "dev-172" "dev-173" "dev-174" "dev-175" "dev-176" "dev-177" "dev-178" "dev-179" "dev-18" "dev-180" "dev-181" "dev-182" "dev-183" "dev-184" "dev-185" "dev-186" "dev-187" "dev-188" "dev-189" "dev-19" "dev-190" "dev-191" "dev-192" "dev-193" "dev-194" "dev-195" "dev-196" "dev-197" "dev-198" "dev-199" "dev-2" "dev-20" "dev-200" "dev-201" "dev-202" "dev-203" "dev-204" "dev-205" "dev-206" "dev-207" "dev-208" "dev-209" "dev-21" "dev-210" "dev-211" "dev-212" "dev-213" "dev-214" "dev-215" "dev-216" "dev-217" "dev-218" "dev-219" "dev-22" "dev-220" "dev-221" "dev-222" "dev-223" "dev-224" "dev-225" "dev-226" "dev-227" "dev-228" "dev-229" "dev-23" "dev-230" "dev-231" "dev-232" "dev-233" "dev-234" "dev-235" "dev-236" "dev-237" "dev-238" "dev-239" "dev-24" "dev-240" "dev-241" "dev-242" "dev-243" "dev-244" "dev-245" "dev-246" "dev-247" "dev-248" "dev-249" "dev-25" "dev-250" "dev-251" "dev-252" "dev-253" "dev-254" "dev-255" "dev-256" "dev-257" "dev-258" "dev-259" "dev-26" "dev-260" "dev-261" "dev-262" "dev-263" "dev-264" "dev-265" "dev-266" "dev-267" "dev-268" "dev-269" "dev-27" "dev-270" "dev-271" "dev-272" "dev-273" "dev-274" "dev-275" "dev-276" "dev-277" "dev-278" "dev-279" "dev-28" "dev-280" "dev-281" "dev-282" "dev-283" "dev-284" "dev-285" "dev-286" "dev-287" "dev-288" "dev-289" "dev-29" "dev-290" "dev-291" "dev-292" "dev-293" "dev-294" "dev-295" "dev-296" "dev-297" "dev-298" "dev-299" "dev-3" "dev-30" "dev-300" "dev-31" "dev-32" "dev-33" "dev-34" "dev-35" "dev-36" "dev-37" "dev-38" "dev-39" "dev-4" "dev-40" "dev-41" "dev-42" "dev-43" "dev-44" "dev-45" "dev-46" "dev-47" "dev-48" "dev-49" "dev-5" "dev-50" "dev-51" "dev-52" "dev-53" "dev-54" "dev-55" "dev-56" "dev-57" "dev-58" "dev-59" "dev-6" "dev-60" "dev-61" "dev-62" "dev-63" "dev-64" "dev-65" "dev-66" "dev-67" "dev-68" "dev-69" "dev-7" "dev-70" "dev-71" "dev-72" "dev-73" "dev-74" "dev-75" "dev-76" "dev-77" "dev-78" "dev-79" "dev-8" "dev-80" "dev-81" "dev-82" "dev-83" "dev-84" "dev-85" "dev-86" "dev-87" "dev-88" "dev-89" "dev-9" "dev-90" "dev-91" "dev-92" "dev-93" "dev-94" "dev-95" "dev-96" "dev-97" "dev-98" "dev-99"]
identity_policies      []
policies               ["default" "dev-1" "dev-10" "dev-100" "dev-101" "dev-102" "dev-103" "dev-104" "dev-105" "dev-106" "dev-107" "dev-108" "dev-109" "dev-11" "dev-110" "dev-111" "dev-112" "dev-113" "dev-114" "dev-115" "dev-116" "dev-117" "dev-118" "dev-119" "dev-12" "dev-120" "dev-121" "dev-122" "dev-123" "dev-124" "dev-125" "dev-126" "dev-127" "dev-128" "dev-129" "dev-13" "dev-130" "dev-131" "dev-132" "dev-133" "dev-134" "dev-135" "dev-136" "dev-137" "dev-138" "dev-139" "dev-14" "dev-140" "dev-141" "dev-142" "dev-143" "dev-144" "dev-145" "dev-146" "dev-147" "dev-148" "dev-149" "dev-15" "dev-150" "dev-151" "dev-152" "dev-153" "dev-154" "dev-155" "dev-156" "dev-157" "dev-158" "dev-159" "dev-16" "dev-160" "dev-161" "dev-162" "dev-163" "dev-164" "dev-165" "dev-166" "dev-167" "dev-168" "dev-169" "dev-17" "dev-170" "dev-171" "dev-172" "dev-173" "dev-174" "dev-175" "dev-176" "dev-177" "dev-178" "dev-179" "dev-18" "dev-180" "dev-181" "dev-182" "dev-183" "dev-184" "dev-185" "dev-186" "dev-187" "dev-188" "dev-189" "dev-19" "dev-190" "dev-191" "dev-192" "dev-193" "dev-194" "dev-195" "dev-196" "dev-197" "dev-198" "dev-199" "dev-2" "dev-20" "dev-200" "dev-201" "dev-202" "dev-203" "dev-204" "dev-205" "dev-206" "dev-207" "dev-208" "dev-209" "dev-21" "dev-210" "dev-211" "dev-212" "dev-213" "dev-214" "dev-215" "dev-216" "dev-217" "dev-218" "dev-219" "dev-22" "dev-220" "dev-221" "dev-222" "dev-223" "dev-224" "dev-225" "dev-226" "dev-227" "dev-228" "dev-229" "dev-23" "dev-230" "dev-231" "dev-232" "dev-233" "dev-234" "dev-235" "dev-236" "dev-237" "dev-238" "dev-239" "dev-24" "dev-240" "dev-241" "dev-242" "dev-243" "dev-244" "dev-245" "dev-246" "dev-247" "dev-248" "dev-249" "dev-25" "dev-250" "dev-251" "dev-252" "dev-253" "dev-254" "dev-255" "dev-256" "dev-257" "dev-258" "dev-259" "dev-26" "dev-260" "dev-261" "dev-262" "dev-263" "dev-264" "dev-265" "dev-266" "dev-267" "dev-268" "dev-269" "dev-27" "dev-270" "dev-271" "dev-272" "dev-273" "dev-274" "dev-275" "dev-276" "dev-277" "dev-278" "dev-279" "dev-28" "dev-280" "dev-281" "dev-282" "dev-283" "dev-284" "dev-285" "dev-286" "dev-287" "dev-288" "dev-289" "dev-29" "dev-290" "dev-291" "dev-292" "dev-293" "dev-294" "dev-295" "dev-296" "dev-297" "dev-298" "dev-299" "dev-3" "dev-30" "dev-300" "dev-31" "dev-32" "dev-33" "dev-34" "dev-35" "dev-36" "dev-37" "dev-38" "dev-39" "dev-4" "dev-40" "dev-41" "dev-42" "dev-43" "dev-44" "dev-45" "dev-46" "dev-47" "dev-48" "dev-49" "dev-5" "dev-50" "dev-51" "dev-52" "dev-53" "dev-54" "dev-55" "dev-56" "dev-57" "dev-58" "dev-59" "dev-6" "dev-60" "dev-61" "dev-62" "dev-63" "dev-64" "dev-65" "dev-66" "dev-67" "dev-68" "dev-69" "dev-7" "dev-70" "dev-71" "dev-72" "dev-73" "dev-74" "dev-75" "dev-76" "dev-77" "dev-78" "dev-79" "dev-8" "dev-80" "dev-81" "dev-82" "dev-83" "dev-84" "dev-85" "dev-86" "dev-87" "dev-88" "dev-89" "dev-9" "dev-90" "dev-91" "dev-92" "dev-93" "dev-94" "dev-95" "dev-96" "dev-97" "dev-98" "dev-99"]
token_meta_username    bob

real	0m2.269s
user	0m0.075s
sys	0m0.070s

$ time vault login -method=ldap username=bob password=password
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  hvs.CAESIExKpeAWV3fxh4mHb9gjObU2204bJDhjSxvdhPfcG4exGh4KHGh2cy5QbERIOHM1TUJKWlhpM0t1ZmtmbFcxTXQ
token_accessor         F0ATCtGE9jjnQDv2IwVdAw0j
token_duration         768h
token_renewable        true
token_policies         ["default" "dev-1" "dev-10" "dev-100" "dev-101" "dev-102" "dev-103" "dev-104" "dev-105" "dev-106" "dev-107" "dev-108" "dev-109" "dev-11" "dev-110" "dev-111" "dev-112" "dev-113" "dev-114" "dev-115" "dev-116" "dev-117" "dev-118" "dev-119" "dev-12" "dev-120" "dev-121" "dev-122" "dev-123" "dev-124" "dev-125" "dev-126" "dev-127" "dev-128" "dev-129" "dev-13" "dev-130" "dev-131" "dev-132" "dev-133" "dev-134" "dev-135" "dev-136" "dev-137" "dev-138" "dev-139" "dev-14" "dev-140" "dev-141" "dev-142" "dev-143" "dev-144" "dev-145" "dev-146" "dev-147" "dev-148" "dev-149" "dev-15" "dev-150" "dev-151" "dev-152" "dev-153" "dev-154" "dev-155" "dev-156" "dev-157" "dev-158" "dev-159" "dev-16" "dev-160" "dev-161" "dev-162" "dev-163" "dev-164" "dev-165" "dev-166" "dev-167" "dev-168" "dev-169" "dev-17" "dev-170" "dev-171" "dev-172" "dev-173" "dev-174" "dev-175" "dev-176" "dev-177" "dev-178" "dev-179" "dev-18" "dev-180" "dev-181" "dev-182" "dev-183" "dev-184" "dev-185" "dev-186" "dev-187" "dev-188" "dev-189" "dev-19" "dev-190" "dev-191" "dev-192" "dev-193" "dev-194" "dev-195" "dev-196" "dev-197" "dev-198" "dev-199" "dev-2" "dev-20" "dev-200" "dev-201" "dev-202" "dev-203" "dev-204" "dev-205" "dev-206" "dev-207" "dev-208" "dev-209" "dev-21" "dev-210" "dev-211" "dev-212" "dev-213" "dev-214" "dev-215" "dev-216" "dev-217" "dev-218" "dev-219" "dev-22" "dev-220" "dev-221" "dev-222" "dev-223" "dev-224" "dev-225" "dev-226" "dev-227" "dev-228" "dev-229" "dev-23" "dev-230" "dev-231" "dev-232" "dev-233" "dev-234" "dev-235" "dev-236" "dev-237" "dev-238" "dev-239" "dev-24" "dev-240" "dev-241" "dev-242" "dev-243" "dev-244" "dev-245" "dev-246" "dev-247" "dev-248" "dev-249" "dev-25" "dev-250" "dev-251" "dev-252" "dev-253" "dev-254" "dev-255" "dev-256" "dev-257" "dev-258" "dev-259" "dev-26" "dev-260" "dev-261" "dev-262" "dev-263" "dev-264" "dev-265" "dev-266" "dev-267" "dev-268" "dev-269" "dev-27" "dev-270" "dev-271" "dev-272" "dev-273" "dev-274" "dev-275" "dev-276" "dev-277" "dev-278" "dev-279" "dev-28" "dev-280" "dev-281" "dev-282" "dev-283" "dev-284" "dev-285" "dev-286" "dev-287" "dev-288" "dev-289" "dev-29" "dev-290" "dev-291" "dev-292" "dev-293" "dev-294" "dev-295" "dev-296" "dev-297" "dev-298" "dev-299" "dev-3" "dev-30" "dev-300" "dev-31" "dev-32" "dev-33" "dev-34" "dev-35" "dev-36" "dev-37" "dev-38" "dev-39" "dev-4" "dev-40" "dev-41" "dev-42" "dev-43" "dev-44" "dev-45" "dev-46" "dev-47" "dev-48" "dev-49" "dev-5" "dev-50" "dev-51" "dev-52" "dev-53" "dev-54" "dev-55" "dev-56" "dev-57" "dev-58" "dev-59" "dev-6" "dev-60" "dev-61" "dev-62" "dev-63" "dev-64" "dev-65" "dev-66" "dev-67" "dev-68" "dev-69" "dev-7" "dev-70" "dev-71" "dev-72" "dev-73" "dev-74" "dev-75" "dev-76" "dev-77" "dev-78" "dev-79" "dev-8" "dev-80" "dev-81" "dev-82" "dev-83" "dev-84" "dev-85" "dev-86" "dev-87" "dev-88" "dev-89" "dev-9" "dev-90" "dev-91" "dev-92" "dev-93" "dev-94" "dev-95" "dev-96" "dev-97" "dev-98" "dev-99"]
identity_policies      []
policies               ["default" "dev-1" "dev-10" "dev-100" "dev-101" "dev-102" "dev-103" "dev-104" "dev-105" "dev-106" "dev-107" "dev-108" "dev-109" "dev-11" "dev-110" "dev-111" "dev-112" "dev-113" "dev-114" "dev-115" "dev-116" "dev-117" "dev-118" "dev-119" "dev-12" "dev-120" "dev-121" "dev-122" "dev-123" "dev-124" "dev-125" "dev-126" "dev-127" "dev-128" "dev-129" "dev-13" "dev-130" "dev-131" "dev-132" "dev-133" "dev-134" "dev-135" "dev-136" "dev-137" "dev-138" "dev-139" "dev-14" "dev-140" "dev-141" "dev-142" "dev-143" "dev-144" "dev-145" "dev-146" "dev-147" "dev-148" "dev-149" "dev-15" "dev-150" "dev-151" "dev-152" "dev-153" "dev-154" "dev-155" "dev-156" "dev-157" "dev-158" "dev-159" "dev-16" "dev-160" "dev-161" "dev-162" "dev-163" "dev-164" "dev-165" "dev-166" "dev-167" "dev-168" "dev-169" "dev-17" "dev-170" "dev-171" "dev-172" "dev-173" "dev-174" "dev-175" "dev-176" "dev-177" "dev-178" "dev-179" "dev-18" "dev-180" "dev-181" "dev-182" "dev-183" "dev-184" "dev-185" "dev-186" "dev-187" "dev-188" "dev-189" "dev-19" "dev-190" "dev-191" "dev-192" "dev-193" "dev-194" "dev-195" "dev-196" "dev-197" "dev-198" "dev-199" "dev-2" "dev-20" "dev-200" "dev-201" "dev-202" "dev-203" "dev-204" "dev-205" "dev-206" "dev-207" "dev-208" "dev-209" "dev-21" "dev-210" "dev-211" "dev-212" "dev-213" "dev-214" "dev-215" "dev-216" "dev-217" "dev-218" "dev-219" "dev-22" "dev-220" "dev-221" "dev-222" "dev-223" "dev-224" "dev-225" "dev-226" "dev-227" "dev-228" "dev-229" "dev-23" "dev-230" "dev-231" "dev-232" "dev-233" "dev-234" "dev-235" "dev-236" "dev-237" "dev-238" "dev-239" "dev-24" "dev-240" "dev-241" "dev-242" "dev-243" "dev-244" "dev-245" "dev-246" "dev-247" "dev-248" "dev-249" "dev-25" "dev-250" "dev-251" "dev-252" "dev-253" "dev-254" "dev-255" "dev-256" "dev-257" "dev-258" "dev-259" "dev-26" "dev-260" "dev-261" "dev-262" "dev-263" "dev-264" "dev-265" "dev-266" "dev-267" "dev-268" "dev-269" "dev-27" "dev-270" "dev-271" "dev-272" "dev-273" "dev-274" "dev-275" "dev-276" "dev-277" "dev-278" "dev-279" "dev-28" "dev-280" "dev-281" "dev-282" "dev-283" "dev-284" "dev-285" "dev-286" "dev-287" "dev-288" "dev-289" "dev-29" "dev-290" "dev-291" "dev-292" "dev-293" "dev-294" "dev-295" "dev-296" "dev-297" "dev-298" "dev-299" "dev-3" "dev-30" "dev-300" "dev-31" "dev-32" "dev-33" "dev-34" "dev-35" "dev-36" "dev-37" "dev-38" "dev-39" "dev-4" "dev-40" "dev-41" "dev-42" "dev-43" "dev-44" "dev-45" "dev-46" "dev-47" "dev-48" "dev-49" "dev-5" "dev-50" "dev-51" "dev-52" "dev-53" "dev-54" "dev-55" "dev-56" "dev-57" "dev-58" "dev-59" "dev-6" "dev-60" "dev-61" "dev-62" "dev-63" "dev-64" "dev-65" "dev-66" "dev-67" "dev-68" "dev-69" "dev-7" "dev-70" "dev-71" "dev-72" "dev-73" "dev-74" "dev-75" "dev-76" "dev-77" "dev-78" "dev-79" "dev-8" "dev-80" "dev-81" "dev-82" "dev-83" "dev-84" "dev-85" "dev-86" "dev-87" "dev-88" "dev-89" "dev-9" "dev-90" "dev-91" "dev-92" "dev-93" "dev-94" "dev-95" "dev-96" "dev-97" "dev-98" "dev-99"]
token_meta_username    bob

real	0m2.221s
user	0m0.060s
sys	0m0.025s
```

Comment (jasonodonnell) [reply sdk/helper/ldaputil/client.go:534]: The existing tests do very little actual testing of these methods. The actual auth method does do a little bit more acceptance testing but token groups are an AD feature and would require a whole new setup to utilize this.

**Review by mpalmi — APPROVED** [review]

**Comment by mpalmi** [comment]
Good stuff! Thanks for looking into the race.

File: sdk/helper/ldaputil/client.go [hunk 2]
```diff
@@ -478,11 +479,15 @@ func sidBytesToString(b []byte) (string, error) {
...
 func (c *Client) performLdapTokenGroupsSearch(cfg *ConfigEntry, conn Connection, userDN string) ([]*ldap.Entry, error) {
+	var wg sync.WaitGroup
+	var lock sync.Mutex
+	taskChan := make(chan string)
+	maxWorkers := 10
+
 	result, err := conn.Search(&ldap.SearchRequest{
-		BaseDN:       userDN,
-		Scope:        ldap.ScopeBaseObject,
-		DerefAliases: ldapDerefAliasMap[cfg.DerefAliases],
```
Comment (mpalmi) [inline sdk/helper/ldaputil/client.go:484]: I think this got lost in the shuffle, perhaps?

Comment (jasonodonnell) [reply sdk/helper/ldaputil/client.go]: Really good catch. Removed during debugging and forgot to add it back. Thanks!

**Review by raymonstah — APPROVED** [review]
Nice improvement. Tagging @jimlambrt in case we want to port these changes to the cap/ldap package as well.

**Review by austingebauer — APPROVED** [review]
Nice improvement! Thanks @jasonodonnell 👍

File: sdk/helper/ldaputil/client.go [hunk 3]
```diff
@@ -478,6 +479,11 @@ func sidBytesToString(b []byte) (string, error) {
 }
 
 func (c *Client) performLdapTokenGroupsSearch(cfg *ConfigEntry, conn Connection, userDN string) ([]*ldap.Entry, error) {
+	var wg sync.WaitGroup
+	var lock sync.Mutex
+	taskChan := make(chan string)
+	maxWorkers := 10
```
[the remark this answers is not in the document]
Comment (jasonodonnell) [reply sdk/helper/ldaputil/client.go:485]: Let's keep it at 10 for now and can dial it back if needed. Interesting pattern though!

**Pull request merged**